### ISSUE TITLE
Get rid of HSNW dependencies

### ISSUE THREAD

**Issue by YuhanLiin** [rust-ml/linfa#164]
If possible, we should replace HSNW dependencies with `linfa-nn`, since both are nearest-neighbour algorithms. Doing so will reduce some duplicated dependencies in our tree.

**Comment by bytesnake** [comment]
also integrate HNSW into `linfa-nn` and make it optionally, discussed on here: https://github.com/rust-ml/linfa/issues/119#issuecomment-841839071

**Comment by YuhanLiin** [comment]
I did some integration work with `rust-cv` a month ago and it looks like we can't integrate HNSW into our interface until GATs are stabilized, due to lifetime issues. If using HSNW over existing `linfa-nn` algorithms is absolutely necessary then we can't do much until GATs land.

**Comment by bytesnake** [comment]
yeah let's integrate `linfa-nn` into places where HSNW is used and add it back to the toolbox later

**Comment by YuhanLiin** [comment]
Closed due to #168
